src/libUtils/UpgradeManager.cpp

```diff
@@ -15,148 +15,387 @@
-}
+        cur = *tok_iter;
 
-/// Implements the Deserialize function inherited from Serializable.
-int SWInfo::Deserialize(const std::vector<unsigned char>& src,
-                        unsigned int offset)
-{
-    LOG_MARKER();
+        if (cur == "browser_download_url")
+        {
+            find = 1;
+            continue;
+        }
+    }
 
-    unsigned int curOffset = offset;
+    string downloadFilePath;
 
-    try
+    for (auto s : downloadFilePaths)
     {
-        m_major = GetNumber<uint32_t>(src, curOffset, sizeof(uint32_t));
-        curOffset += sizeof(uint32_t);
-        m_minor = GetNumber<uint32_t>(src, curOffset, sizeof(uint32_t));
-        curOffset += sizeof(uint32_t);
-        m_fix = GetNumber<uint32_t>(src, curOffset, sizeof(uint32_t));
-        curOffset += sizeof(uint32_t);
-        m_upgradeDS = GetNumber<uint64_t>(src, curOffset, sizeof(uint64_t));
-        curOffset += sizeof(uint64_t);
-        m_commit = GetNumber<uint32_t>(src, curOffset, sizeof(uint32_t));
-        curOffset += sizeof(uint32_t);
+        if (string::npos != s.rfind(fileTail))
+        {
+            downloadFilePath = s;
+            break;
+        }
     }
-    catch (const std::exception& e)
+
+    string fileName = downloadFilePath.substr(downloadFilePath.rfind('/') + 1);
+
+    /// Get the redirection url (if applicable)
+    curl_easy_reset(m_curl);
+    curl_easy_setopt(m_curl, CURLOPT_URL, downloadFilePath.data());
+    res = curl_easy_perform(m_curl);
+
+    if (res != CURLE_OK)
     {
         LOG_GENERAL(WARNING,
-                    "Error with SWInfo::Deserialize." << ' ' << e.what());
-        return -1;
+                    "curl_easy_perform() failed: " << curl_easy_strerror(res));
+        return "";
     }
 
-    return 0;
-}
+    long response_code;
+    res = curl_easy_getinfo(m_curl, CURLINFO_RESPONSE_CODE, &response_code);
 
-/// Less-than comparison operator.
-bool SWInfo::operator<(const SWInfo& r) const
-{
-    return tie(m_major, m_minor, m_fix, m_upgradeDS, m_commit)
-        < tie(r.m_major, r.m_minor, r.m_fix, r.m_upgradeDS, r.m_commit);
-}
+    if ((res != CURLE_OK) || ((response_code / 100) == 3))
+    {
+        char* location;
+        curl_easy_getinfo(m_curl, CURLINFO_REDIRECT_URL, &location);
 
-/// Greater-than comparison operator.
-bool SWInfo::operator>(const SWInfo& r) const { return r < *this; }
+        if ((res == CURLE_OK) && location)
+        {
+            downloadFilePath = location;
+        }
+    }
 
-/// Equality operator.
-bool SWInfo::operator==(const SWInfo& r) const
-{
-    return tie(m_major, m_minor, m_fix, m_upgradeDS, m_commit)
-        == tie(r.m_major, r.m_minor, r.m_fix, r.m_upgradeDS, r.m_commit);
-}
+    /// Download the file
+    curl_easy_reset(m_curl);
+    curl_easy_setopt(m_curl, CURLOPT_URL, downloadFilePath.data());
+    curl_easy_setopt(m_curl, CURLOPT_VERBOSE, 1L);
+    curl_easy_setopt(m_curl, CURLOPT_NOPROGRESS, 1L);
+    FILE* file = fopen(fileName.data(), "wb");
```

Comment:
RAII can make this more maintainable, see https://github.com/Zilliqa/Issues/issues/172